[Order Form] Improve custom amounts forms

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: WOOMOB-139
Closes: WOOMOB-138
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR improves the Custom Amount forms, fixing two specific bugs and simplifying the code.

Bugs fixed:
 - WOOMOB-139: the `Add custom amount` confirmation button gets hidden under the keyboard on iPad and can't be scrolled into view.
 - WOOMOB-138: the percentage entry field can truncate for long numbers entered, or at large dynamic type sizes.

I've also made the percentage entry style match the amount entry more closely, and both match the shipping line entry.

I tried to format the percentage properly based on locale, but couldn't get that working correctly in a reasonable amount of time unfortunately. We'll come back to it, it's not a regression.

It would be nice to make the amount fields a single reusable view, but they're different enough that I didn't want to handle that just here. At least they match visually now!

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Launch the app and navigate to orders
2. Add a new order, and add a custom amount – observe that the button to add it is floated above the keyboard, and all other items are scrollable.
3. Add a shipping line – observe that the style is the same
4. Edit the custom amount as a percentage – observe that it's styled correctly

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I've tested on an iPad Air running iOS 17.7.2, and an iPhone 15 Pro Max running iOS 18.4

It's worth testing with different locales and dynamic type sizes.


## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/4aac6766-da3c-45ad-953f-07268195c651

https://github.com/user-attachments/assets/2bc536ed-61a9-464d-b466-3b64501cda6f



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
Allow the add amount button to avoid the keyboard

joshheald committed:
Update custom amount view to use shipping style

joshheald committed:
Update custom percentage view to new style

joshheald committed:
Update percentage text field to match amount field

Also fixes WOOMOB-138 by removing the constrained width on the percentage entry.

joshheald committed:
Remove keyboard dismiss workaround

This is no longer required now the buttons always clear the keyboard.

It appeared to also be interfering with proper use of the form

joshheald committed:
Ensure that taps on the percentage field focus it

joshheald committed:
Localize formatting of entered percentages

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Add custom amount improvements to 22.3 release

joshheald committed:
Remove broken attempt to localise percentages

This was close to working… but if you typed 40.001, the two zeroes after the decimal wouldn’t show up until the 1 was typed. We can handle this separately.

joshheald committed:
Merge branch 'trunk' into bug/woomob-139-order-form-submit-button-on-addedit-custom-amount-form-is

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountPercentageView.swift:
```diff
@@ -56,27 +55,33 @@ private extension AddCustomAmountPercentageView {
+                .opacity(0)
 
-                Text("%")
+                Text(text + "%")
```
I'm not sure if the percentage symbol shows differently in RTL locales, but perhaps is worth to cover the case:
```
Text(String.localizedStringWithFormat("%@%%", text))
```

joshheald replied:
Since we need to address it anyway, better to do it with a formatter. A good time will be when we update the calculation approach as discussed yesterday. No need to make a confusing localized string for now; it's been this way for ages without hearing about it anyway.

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountPercentageView.swift:
```diff
@@ -4,32 +4,27 @@ struct AddCustomAmountPercentageView: View {
+        Group {
+            VStack(alignment: .leading) {
+                LabeledContent {
+                    Text("(\(viewModel.baseAmountForPercentageString))")
```
This one is pre-existing:
```suggestion
                    Text(viewModel.baseAmountForPercentageString)
```

iamgabrielma reviewed (approved):
LGTM!

> It's worth testing with different locales and dynamic type sizes.

One of the bits that actually works well in larger sizes 😅 we tend to cut off the text pretty often all across the board.

![Simulator Screenshot - iPhone 16 - 2025-04-30 at 09 17 59](https://github.com/user-attachments/assets/53b9f918-92ff-43c5-892e-833bddfb6eca)

joshheald committed:
Update WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomAmounts/AddCustomAmountPercentageView.swift

Co-authored-by: Gabriel Maldonado <gabriel.maldonado@automattic.com>

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/243">22.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
